### PR TITLE
fix: [PL-48009]: allow - (hyphen) in identifier

### PR DESCRIPTION
**Pull request by namanharness**
Changelist: allow - (hyphen) in identifier
The ask is from Skillsoft to allow `-` in Secret identifier.

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- StyleLint: `retrigger stylelint`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- ImageSnapshot `retrigger ImageSnapshot`

### CODE REVIEW

**Comment by bot2-harness** [comment]
Image snapshots failed please check https://console.cloud.google.com/storage/browser/ui-team-bucket/uicore-2361

File: packages/uicore/src/components/InputWithIdentifier/InputWithIdentifier.tsx [hunk 1]
```diff
@@ -57,12 +57,24 @@ export interface InputWithIdentifierProps {
...
 // https://harness.atlassian.net/wiki/spaces/CDNG/pages/736200458/Entity+Identifier
-export function getIdentifierFromName(str: string): string {
+export function getIdentifierFromName(str: string, allowHyphen: boolean): string {
   if (isNil(str)) return ''
 
+  if (allowHyphen) {
+    return str
+      .trim()
+      .replace(/^[0-9-$]*/, '') // remove starting digits, dashes and $
+      .replace(/[^0-9a-zA-Z_$ -]/g, '') // remove special chars except _ and $ and -
```
Comment (namanharness) [inline packages/uicore/src/components/InputWithIdentifier/InputWithIdentifier.tsx:75]: change in regex in this line only where we have added `-` in addition to `_$`

File: packages/uicore/src/components/InputWithIdentifier/InputWithIdentifier.tsx [hunk 2]
```diff
@@ -57,12 +57,24 @@ export interface InputWithIdentifierProps {
...
+  /**
+   * Allow hyphen in identifier
+   * @default false
+   */
+  allowHyphen?: boolean
 }
 
 // https://harness.atlassian.net/wiki/spaces/CDNG/pages/736200458/Entity+Identifier
-export function getIdentifierFromName(str: string): string {
+export function getIdentifierFromName(str: string, allowHyphen: boolean): string {
```
Comment (mayankVermaHarness) [inline packages/uicore/src/components/InputWithIdentifier/InputWithIdentifier.tsx:68]: Please add jest test cases for this method.

Let's also ensure to update this doc: https://harness.atlassian.net/wiki/spaces/CDNG/pages/736200458/Entity+Identifier

File: packages/uicore/src/components/FormikForm/__tests__/FormikForm.test.tsx [hunk 3]
```diff
@@ -573,4 +574,28 @@ describe('<FormInput.KVTagInput />', () => {
 
     await waitFor(() => expect(onSubmit).toBeCalledWith({ tags: ['t1', 't2', 't3', 't4'] }, expect.anything()))
   })
+  test('test getIdentifierFromName() function', () => {
```
Comment (mayankVermaHarness) [inline packages/uicore/src/components/FormikForm/__tests__/FormikForm.test.tsx:577]: Pretty comprehensive! Good work!

**Comment by bot2-harness** [comment]
Image snapshots failed please check https://console.cloud.google.com/storage/browser/ui-team-bucket/uicore-2362

**Review by mayankVermaHarness — APPROVED** [review]

File: packages/uicore/src/components/FormikForm/__tests__/FormikForm.test.tsx (hunk 3 above)
Comment (chetanvekariya-harness) [reply packages/uicore/src/components/FormikForm/__tests__/FormikForm.test.tsx:577]: Yes, good work. But, we should also add one test case for NameWithIdentifier component.

Comment (namanharness) [reply packages/uicore/src/components/FormikForm/__tests__/FormikForm.test.tsx:577]: i think you mean `InputWithIdentifier` component.
image

that will be tested by default from `harness-core-ui` after this change

Comment (chetanvekariya-harness) [reply packages/uicore/src/components/FormikForm/__tests__/FormikForm.test.tsx:577]: Yeah, I think that's the one. Component should have its own test cases. Ideally, we do not need to test in harness-core-ui. But, anyway that gets tested everywhere along with other things.
If I create new micro-frontend tomorrow, use this component and if somehow flow breaks, how would I know? So, it is important to have tests for all use cases in a dedicated test file of that component.

**Comment by bot2-harness** [comment]
Image snapshots failed please check https://console.cloud.google.com/storage/browser/ui-team-bucket/uicore-2367

File: packages/uicore/src/components/InputWithIdentifier/__tests__/InputWithIdentifier.test.tsx [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+describe('Test InputWithIdentifier', () => {
+  function addNameAndAssertModifiedIdentifier(
+    name: string,
+    identifier: string,
+    input: HTMLInputElement,
+    idClass: HTMLInputElement
+  ) {
+    userEvent.type(input, name)
+    expect(input).toHaveValue(name)
+    expect(within(idClass).getByText(identifier)).toBeInTheDocument()
```
Comment (chetanvekariya-harness) [inline packages/uicore/src/components/InputWithIdentifier/__tests__/InputWithIdentifier.test.tsx:41]: nit: usage of `screen.getByText` would have been fine.

**Review by chetanvekariya-harness — APPROVED** [review]

**Comment by bot2-harness** [comment]
Image snapshots failed please check https://console.cloud.google.com/storage/browser/ui-team-bucket/uicore-2368

**Review by mayankVermaHarness — APPROVED** [review]

**Pull request merged**